### PR TITLE
[Tubepatrol] Add new extractor

### PR DESCRIPTION
**Pull request by parmjitv**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x ] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x ] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Added a new extractor for yet another porn site [Tubepatrol.sex]

This site largely serves as a mashup of video content from other providers - Pornhub, xHamster etc.

The basic extractor function has been provided for the Video ID, Display ID, Title and URL parameters only.

Cheers and thank you!

Parmjit V.

### CODE REVIEW

**Comment by dstftw** [comment]
Both test cases are borfos.com generic embeds thus:
1. This extractor should be removed.
2. Extractor for `https://borfos.com/embed/\d+` should be added.
3. Generic extractor should be modified to match such embeds.

**Comment by parmjitv** [comment]
Thanks for the prompt review.  Can you please elaborate?

The [borfos.com] embed URLs do not provide the required extractor metadata, namely the video title text and display ID.

As it is, extracting from [tubepatrol.sex] gives us access to this required metadata, and will allow us to also composite some optional fields in a subsequent commit - specifically 'categories' and 'view_count'.  The current implementation also provides for a fallback video URL.

Cheers,

Parmjit

**Comment by dstftw** [comment]
As already said generic embed must be processed by generic extractor. As per additional metadata it can be added using `url_transparent` extraction entity result, but I wouldn't bother with that since it's not feasible to write extractors for all sites embedding this kind of videos.

File: youtube_dl/extractor/extractors.py [hunk 1]
```diff
@@ -1309,4 +1310,4 @@
 from .zapiks import ZapiksIE
 from .zaq1 import Zaq1IE
 from .zdf import ZDFIE, ZDFChannelIE
-from .zingmp3 import ZingMp3IE
+from .zingmp3 import ZingMp3IE
```
Comment (dstftw) [inline youtube_dl/extractor/extractors.py]: Don't touch unrelated stuff.

File: youtube_dl/extractor/tubepatrol.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
+# coding: utf-8
+from __future__ import unicode_literals
+from ..utils import ExtractorError
+from .common import InfoExtractor
+import re
+
+
+class TubepatrolIE(InfoExtractor):
+    # i.e. http://tubepatrol.sex/to/767066/plump-asian-loves-fucking-and-sucking.html
```
Comment (dstftw) [inline youtube_dl/extractor/tubepatrol.py]: It's already clear from test URLs.

File: youtube_dl/extractor/tubepatrol.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                'id': '3934608',
+                'display_id': 'ad4x-video-dp-de-kelly-lee-trailer-hd-porn-quebec',
+                'ext': 'flv',
+                'title': 'AD4X Video - DP De Kelly Lee Trailer HD - Porn Quebec',
+            },
+        },
+    ]
+
+    def _real_extract(self, url):
+        # Basic extractor implementation - Video ID, Display ID, Title, URL
```
Comment (dstftw) [inline youtube_dl/extractor/tubepatrol.py]: Remove all pointless comments.

File: youtube_dl/extractor/tubepatrol.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+        # IDs
+
+        mobj = re.match(self._VALID_URL, url)
+        try:
+            video_id = mobj.group('id')
+            display_id = mobj.group('display_id')
+        except Exception as inst:
```
Comment (dstftw) [inline youtube_dl/extractor/tubepatrol.py]: That's impossible outcome.

File: youtube_dl/extractor/tubepatrol.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        # get the webpage source code
+        webpage = self._download_webpage(url, video_id)
+
+        # Title
+
+        # first try the generic header text
+        video_title = self._html_search_regex(r'<h1>(.+?)</h1>', webpage, 'video_title', default=None)
+        if video_title is None:
+            # fallback to the link text provided for embed
+            video_title = self._html_search_regex(r'<a\shref="%s">(.+?)</a>' % url, webpage, 'video_title', default=None)
```
Comment (dstftw) [inline youtube_dl/extractor/tubepatrol.py]: 1. 53-56: single `_html_search_regex` with multiple regexes.
2. -`default=None`.

File: youtube_dl/extractor/tubepatrol.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        if video_title is None:
+            raise ExtractorError('%s said: %s' % (self.IE_NAME, "The video title could not be extracted"), expected=True)
```
Comment (dstftw) [inline youtube_dl/extractor/tubepatrol.py]: Remove.

File: youtube_dl/extractor/tubepatrol.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+        # the URL for the video file is contained in a seperate link as:
+        # https://borfos.com/kt_player/player.php?id=<video_id>
+        flashvars_webpage = self._download_webpage('https://borfos.com/kt_player/player.php?id=%s' % video_id, video_id)
+        flashvars_data = self._search_regex(r'(?s)flashvars\s*=\s*({.+?})', flashvars_webpage, 'flashvars_data', default=None)
+        if flashvars_data is None:
+            raise ExtractorError('%s said: %s' % (self.IE_NAME, "The flash player data could not be extracted"), expected=True)
```
Comment (dstftw) [inline youtube_dl/extractor/tubepatrol.py]: 63-65: see review for `video_title`.

File: youtube_dl/extractor/tubepatrol.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        # this is done because a bunch of extraneous fields in the flash data contain wonky characters
+        # that screw up the call to _parse_json() and we do not care for these fields anyway, so ...
+
+        # first try the generic url
+        video_url = self._search_regex(r'(?s)video_url:\s"(.+?)",\s*', flashvars_data, 'video_url', default=None)
+        if video_url is None:
+            # fallback to the HTML5 url
+            video_url = self._search_regex(r'(?s)video_html5_url:\s"(.+?)",\s*', flashvars_data, 'video_url', default=None)
+        if video_url is None:
+            raise ExtractorError('%s said: %s' % (self.IE_NAME, "The video URL could not be extracted"), expected=True)
```
Comment (dstftw) [inline youtube_dl/extractor/tubepatrol.py]: Same.

**Comment by parmjitv** [comment]
OK, thank you for the valuable feedback.  Not too bad for a first pull request :)

Seeing as this new extractor appears to be redundant, shall we close this pull request?  I can update the generic extractor as you request and submit a new PR for review at a later date.

Yours,

Parmjit
